File: src/formatters/PAM_Formatter.java
package formatters;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.RandomAccessFile;
import java.nio.charset.StandardCharsets;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import image.Image;
import image.Layer;
import image.Pixel;

public class PAM_Formatter extends Formatter {

	@SuppressWarnings("serial")
	public static class PAMFormatNotSupported extends Formatter.FormatNotSupported {
		public PAMFormatNotSupported(int depth) {
			message = "PAM file with depth parameter" + depth + " not supported.";
		}

		@Override
		public String getMessage() {
			return this.message;
		}

	}
	private long width;
	private long height;
	private int depth = 4;

	public PAM_Formatter(String filePath) {
		super(filePath);
	}

	@Override
	protected Pixel getPixel(RandomAccessFile reader) throws IOException {
		byte[] buffer = new byte[4];

		if (this.depth == 4) {
			reader.read(buffer);
		}
		else {
			reader.read(buffer, 0, 3);
			buffer[3] = (byte) 0xFF;
		}


		return new Pixel(buffer[0], buffer[1], buffer[2], buffer[3]);
	}

	@Override
	protected void putPixel(RandomAccessFile writer, Pixel pixel) throws IOException {
		byte[] buffer = { pixel.red, pixel.green, pixel.blue, pixel.alpha };
		writer.write(buffer);
	}

	@Override
	public void read(Image image) throws IOException, PAMFormatNotSupported {
		RandomAccessFile reader = new RandomAccessFile(this.filePath, "r");

		String str;
		Pattern pattern = Pattern.compile(".* (.*)");
		Matcher matcher;
		reader.readLine(); // P7\n

		str = reader.readLine(); //WIDTH xxxx\n
		matcher = pattern.matcher(str);
		if (matcher.matches())
			this.width = Long.parseLong(matcher.group(1));

		str = reader.readLine(); //HEIGHT xxxx\n
		matcher = pattern.matcher(str);
		if (matcher.matches())
		this.height = Long.parseLong(matcher.group(1));

		str = reader.readLine(); //DEPTH xxxx\n
		matcher = pattern.matcher(str);
		if (matcher.matches())
			this.depth = Integer.parseInt(matcher.group(1));

		if (depth < 3) {
			reader.close();
			throw new PAMFormatNotSupported(depth);

		}

		reader.readLine();
		reader.readLine();
		reader.readLine();

		Layer layer = new Layer(image, this.filePath, this.height, this.width, true, true, Layer.MAX_OPACITY, Layer.DEFAULT_COORDINTES, image.getLayerNum());
		Pixel[][] matrix = layer.getMatrix();
		for (int i = 0; i < layer.getHeight(); ++i)
			for (int j = 0; j < layer.getWidth(); ++j)
				matrix[i][j] = this.getPixel(reader);
		reader.close();
		image.addLayer(layer);

	}

	@Override
	public void write(Image image) throws IOException {
		final Layer layer = image.getProject();

		OutputStreamWriter headerWriter = new OutputStreamWriter(new FileOutputStream(this.filePath), StandardCharsets.UTF_8);

		StringBuilder str = new StringBuilder("P7\n");
		str.append("WIDTH " + layer.getWidth() + "\n");
		str.append("HEIGHT " + layer.getHeight() + "\n");
		str.append("DEPTH " + this.depth + "\n");
		str.append("MAXVAL 255\n");
		str.append("TUPLTYPE RGB_ALPHA\n");
		str.append("ENDHDR\n");
		headerWriter.write(str.toString());
		headerWriter.close();

		RandomAccessFile writer = new RandomAccessFile(this.filePath, "rw");
		writer.seek(writer.length());

		final Pixel[][] matrix = layer.getMatrix();
		for (int i = 0; i < layer.getHeight(); ++i)
			for (int j = 0; j < layer.getWidth(); ++j)
				this.putPixel(writer, matrix[i][j]);
		writer.close();

	}



}

File: C++ Project/src/PAM_Formatter.cpp
#include "PAM_Formatter.h"

#include "Image.h"
#include "Utilities.h"

#include <fstream>
#include "exceptions.h"

Pixel PAM_Formatter::get_pixel(std::ifstream & ifs) const {
	Dword dword;
	if (depth == 4) {										// RGBA 
		ifs.read(reinterpret_cast<char*>(&dword), 4);		// ABGR
		dword =	swap_bytes(dword);							// RGBA
	}
	else {													// RGB
		ifs.read(reinterpret_cast<char*>(&dword), 3);		// 0BGR
		dword = swap_bytes(dword);							// RGB0
		dword |= 0xFFU;										// RGBA
	}
	return Pixel(dword);
}
void PAM_Formatter::put_pixel(std::ofstream & ofs, const Pixel & pixel) const {
	Dword dword = static_cast<Dword>(pixel);					// RGBA	
	dword = swap_bytes(dword);									// ABGR
	ofs.write(reinterpret_cast<char*>(&dword), sizeof(Pixel));  // puts : RGBA 
}
void PAM_Formatter::read(Image& image) {
	std::ifstream ifs(this->file_path, std::ios::in | std::ios::binary);
	if (!ifs.is_open())
		throw File_Opening_Exception(file_path);

	std::string dummy;
	std::getline(ifs, dummy, ' ');			// P7\nWIDTH_
	ifs >> width;
	std::getline(ifs, dummy, ' ');			// \nHEIGHT_
	ifs >> height;
	std::getline(ifs, dummy,' ');			// \nDEPTH
	ifs >> depth;
	if (depth < 3)
		throw PAM_Format_Not_Supported(depth);			// YET!!
	std::getline(ifs, dummy);				// \n
	std::getline(ifs, dummy);				// MAXVAL
	std::getline(ifs, dummy);				// TUPLTYPE
	std::getline(ifs, dummy);				// ENDHDR

	Layer layer(file_path, height, width);
	for (auto& it : layer.get_matrix())
		for (auto& jt : it)
			jt = get_pixel(ifs);
	ifs.close();

	image.add_layer(layer);
}
void PAM_Formatter::write(const Image& image) const {
	std::ofstream ofs(this->file_path, std::ios::out | std::ios::binary | std::ios::trunc);
	if (!ofs.is_open())
		throw File_Opening_Exception(file_path);

	const Layer& layer = image.get_project();

	ofs << "P7\n";
	ofs << "WIDTH " << layer.get_width() << '\n';
	ofs << "HEIGHT " << layer.get_height() << '\n';
	ofs << "DEPTH " << depth << '\n';
	ofs << "MAXVAL 255\n";
	ofs << "TUPLTYPE RGB_ALPHA\n";
	ofs << "ENDHDR\n";

	for (auto it : layer.get_matrix())
		for (auto jt : it)
			put_pixel(ofs, jt);

	ofs.close();
}
